daemon/src/db/failed.rs

```diff
@@ -0,0 +1,556 @@
+            .add_opening_fee(cfd.opening_fee)
+            .add_funding_fee(initial_funding_fee);
+
+        Fees::new(fee_account.balance())
+    };
+
+    let query_result = sqlx::query!(
+        r#"
+        INSERT INTO failed_cfds
+        (
+            uuid,
+            position,
+            initial_price,
+            taker_leverage,
+            n_contracts,
+            counterparty_network_identity,
+            role,
+            fees,
+            kind
+        )
+        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+        "#,
+        cfd.id,
+        cfd.position,
+        cfd.initial_price,
+        cfd.taker_leverage,
+        n_contracts,
+        cfd.counterparty_network_identity,
+        cfd.role,
+        fees,
+        kind,
+    )
+    .execute(&mut *conn)
+    .await?;
+
+    if query_result.rows_affected() != 1 {
+        anyhow::bail!("failed to insert into failed_cfds");
+    }
+
+    Ok(())
+}
+
+async fn insert_event_log(
+    conn: &mut Transaction<'_, Sqlite>,
+    id: OrderId,
+    event_log: EventLog,
+) -> Result<()> {
+    for EventLogEntry { name, created_at } in event_log.0.iter() {
+        let query_result = sqlx::query!(
+            r#"
+            INSERT INTO event_log_failed (
+                cfd_id,
+                name,
+                created_at
+            )
+            VALUES
+            (
+                (SELECT id FROM failed_cfds WHERE failed_cfds.uuid = $1),
+                $2, $3
+            )
+            "#,
+            id,
+            name,
+            created_at
+        )
+        .execute(&mut *conn)
+        .await?;
+
+        if query_result.rows_affected() != 1 {
+            anyhow::bail!("failed to insert into event_log_failed");
+        }
+    }
+
+    Ok(())
+}
+
+/// Obtain the time at which the failed CFD was created, according to
+/// the `event_log_failed` table.
+///
+/// A CFD can fail either because:
+///
+/// - the offer was rejected; or
+/// - contract setup finished unsuccessfully.
+///
+/// Therefore we can take the creation timestamp of either the
+/// `EventKind::OfferRejected` variant or the
+/// `EventKind::ContractSetupFailed` variant.
+///
+/// We choose to not depend on `EventKind::ContractSetupStarted`
+/// because it's impossible to automatically ensure that we will
+/// always emit such an event before emitting the
+/// `EventKind::ContractSetupFailed`. The slight inaccuracy is
+/// preferred over the possibility of running into bugs.
+async fn load_creation_timestamp(
+    conn: &mut PoolConnection<Sqlite>,
+    id: OrderId,
+) -> Result<Timestamp> {
+    let row = sqlx::query!(
+        r#"
+        SELECT
+            event_log_failed.created_at
+        FROM
+            event_log_failed
+        JOIN
+            failed_cfds on failed_cfds.id = event_log_failed.cfd_id
+        WHERE
+            failed_cfds.uuid = $1 AND (event_log_failed.name = $2 OR event_log_failed.name = $3)
```

Comment:
🛠 I read the comment, but I am not sure why we cannot always `select` the first event in the log. I would prefer that over the current solution. I think an `ASC LIMIT 1` should do the job or am I missing something?

Reply:
Fair enough, I did not consider taking the first event! I'll try that instead.